File: leetcode_C++/p142.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution
{
public:
    ListNode *detectCycle(ListNode *head)
    {
        // 快慢指针
        // 当快慢指针相等时候，或者快指针到顶点时候结束
        // 指针相等时候，fast在圈里走了slow同样的步数，则
        ListNode *fast = head;
        ListNode *slow = head;
        while (true)
        {
            if (fast == nullptr || fast->next == nullptr)
            {
                return nullptr;
            }
            fast = fast->next->next;
            slow = slow->next;
            if (fast == slow)
            {
                break;
            }
        }

        fast = head;
        while (slow != fast)
        {
            fast = fast->next;
            slow = slow->next;
        }
        return fast;
    }
};

File: leetcode_py/p142.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        fast = last = head
        while True:
            if not fast or not fast.next:
                return None
            if fast == last:
                break
            fast = fast.next.next
            last = last.next
        fast = head
        while fast != last:
            fast = fast.next
            last = last.next
            
        return fast